{
 "metadata": {
  "name": "",
  "signature": "sha256:37bea714e9eed1c64be91bf94835d28c660fbba0b65e721ad626b483baf2c048"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Inferring Transcription Factor Activities by Combining Binding Information with Gene Expression Profiles"
     ]
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "A Model and Analysis in Progress"
     ]
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "30th May 2014"
     ]
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "M. Arif Rahman and Neil D. Lawrence"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "In this notebook we design a covariance function for reconstructing transcription factor activities given gene expression profiles and a connectivity matrix (binding data) between genes and transcription factors. Our modelling framework builds on ideas in [Sanguinetti et al (2006)](http://bioinformatics.oxfordjournals.org/content/22/14/1753.short) who used a linear-Gaussian statespace modelling framework to infer the transcription factor activity of a group of genes. \n",
      "\n",
      "We note that the linear Gaussian model is equivalent to a Gaussian process with a particular covariance function. We therefore build a model directly from the Gaussian process perspective to achieve the same effect. We introduce a computational trick, based on  judicious application of singluar value decomposition, to enable us to efficiently fit the Gaussian process in a reduced 'TF activity' space. "
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import numpy as np\n",
      "import pylab as pb\n",
      "import GPy\n",
      "import scipy.io\n",
      "import matplotlib.pyplot as plt\n",
      "\n",
      "%matplotlib inline"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "First we load in the classic [Spellman et al (1998)](http://www.molbiolcell.org/content/9/12/3273.full) Yeast Cell Cycle data set. The cdc15 time series data has 23 time points. We can load this gene expression data in with GPy."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "data = GPy.util.datasets.spellman_yeast_cdc15()\n",
      "Y = data['Y'].fillna(0) # Replace missing values with zero following Sanguinetti et al.\n",
      "t = data['t']\n",
      "print data['info'], data['details']"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Time series of synchronized yeast cells from the CDC-15 experiment of Spellman et al (1998). Two colour spotted cDNA array data set of a series of experiments to identify which genes in Yeast are cell cycle regulated.\n"
       ]
      }
     ],
     "prompt_number": 2
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "More details on the data are given in [this notebook](../datasets/spellmanYeastData.ipynb)."
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "We can make a simple helper function to plot genes from the data set (which are provided as a `pandas` array)."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def plot_gene(gene_name='YER124C'):\n",
      "    plt.plot(data['t'], data['Y'][gene_name], 'rx')\n",
      "    plt.title('Gene: ' + gene_name)\n",
      "    plt.xlabel('time/minutes')\n",
      "plot_gene('YER124C')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "display_data",
       "png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAEZCAYAAABsPmXUAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAHeBJREFUeJzt3Xt0VNXdxvFnImAQgoMrAlkGISKFEEIShRBixUEkIAjY\nUkRq8QJCVbQqS7y8iAYvIKAFpYXlUkEF7cK6iuKLELQ6IBGCSLiq5WIiSElFIBC5BrLfP+bNlCEX\nkswMM7Pz/ayVlcyZmXP2ztGHnd/ZZ4/DGGMEALBKVKgbAAAIPMIdACxEuAOAhQh3ALAQ4Q4AFiLc\nAcBChDsAWIhwR8Dk5OTI5XKpRYsWuvTSS5WRkaE5c+aEtE1bt26V0+nU9u3bfbb37t1b//M//yNJ\nioqKUtOmTRUTE+P9evHFFyVJ2dnZatiwoWJiYtS6dWvdcccdWr9+vXc/paWl+t3vfqeEhARFRUVp\nxYoVPseZPn26kpOT5XQ61a9fPy1YsKDSdq5YsUJRUVGaOHGiz/ZDhw5p7NixSkxMVLNmzZSYmKjs\n7GwdPXrU798N7Ea4IyBee+013XbbberWrZvcbrf27Nmjl19+WTk5OTp58mTI2pWUlKRHHnlEo0aN\n8m574403tHfvXmVnZ3u3bdq0SSUlJd6vRx55RJLkcDg0fPhwlZSUaOPGjYqOjtajjz7qc4yePXtq\nwYIFatWqlRwOR4U2zJ8/X/v379fEiRP14IMP6ttvv/V5vrS0VA8++KAyMjJ83n/s2DF17dpV33zz\njSZNmqSff/5ZH3zwgXbt2qUdO3YE4tcDmxnAT4cPHzbNmzc3U6ZMqfZ1paWlZuHChaZXr14mJSXF\nvP766+bEiRPGGGM+//xzc9lll5lXX33VJCQkmMzMTPPxxx/X6L3nUlpaalJTU81f//pXU1RUZGJj\nY83q1au9zzscDrNz585K3/v000+bP/zhD97Hq1evNlFRUebo0aMVXhsfH29WrFhRbVvuvvtu89hj\nj/lsmzJlinnsscfMnXfeaZ588knv9ueee840a9bMnDx5skb9BM7UINT/uCDybdmyRcXFxRo0aFC1\nr5s9e7Y+/PBDzZo1S9HR0br77rtVVlam0aNHS5J++uknbdiwQXl5efrwww81ZswY7d69u0bvbd68\nuZYsWaLMzMwKx23QoIHmzZun3r1764MPPtCIESOUkZHh8xpTg1U49u3bp9dff12pqalq3LhxjX43\nZzp9+rTWrl2rP/3pT95tP/zwg+bNm6f169dr7NixPq9fuXKlsrKy1LBhw1ofC2DkDr8tXLjQNGvW\nzGdbjx49jNPpNI0bNzZffPGFMcaYzMxMk5ub633NokWLTP/+/Y0xnpH7BRdcYPbt22eM8Yy2mzZt\nar777rtzvremxo8fb+Lj482xY8d8tjscDtOsWTPjdDq9X8uXLzfGeEbujRo1Mk6n00RFRZnLLrvM\n28aznWvk/sQTT5irr77anD592rtt0KBB5r333jPGGHPnnXeaiRMnep/r1KmT+fOf/1yrPgLlGLnD\nb61bt1ZJSYk2b96s5ORkSdKXX37pfa6srExHjhzR6tWrNWDAAO/7jDE+Nea4uDjFxsZK8oy2Y2Nj\ntWfPHsXHx5/zvTXRqVMntW3bVtHR0RWey8/P1xVXXFHp+4YNG6a3335bBQUFuuWWWzR16lRNnz69\nVsd+5ZVXtHDhQq1atUpRUZ5LXR999JF++eUXDR061Nsnc8ZfEK1bt9aXX36phx9+uFbHAiQuqCIA\nOnfuLKfTqcWLF1f5miZNmqh79+7KycnRwYMHdfDgQRUXF+vgwYPn3L8/7/WXw+HwBm5CQoLefPNN\n/eUvf9H3339f433MnTtXL774oj799FPFxcV5t3/22Wdat26d4uLiFBcXp/fee08zZ87Ub37zG0me\nC7XLly9XaWlpYDuFeoFwh99iYmL0wgsvaMaMGXr00Ue1ZcsWlZWVacOGDTpy5Ij3dSNGjNBTTz2l\n9evXq6ysTHv27NHy5ctrdAx/3lsTpoqa+9nbk5KSdNNNN/mM3E+cOKHjx49X+FmS3nnnHU2YMEE5\nOTlKSEjw2dezzz6r7du3a+PGjdqwYYMGDRqkMWPGaN68eZKkcePGqWXLlurXr5/ef/99nThxQtu3\nb9eYMWO0adOmgPQb9iLcERBjxozRggULtHbtWvXq1UutWrXSPffco2nTpqlHjx6SpNGjR2vkyJF6\n6qmndMkll6hPnz7atm2bdx/VlVnO9d6YmBjl5uZW20aHw1HlMVJSUnzmuY8bN67K94wfP15vv/22\nfvrpJ0lShw4ddNFFF+nf//63+vbtqyZNmmjXrl2SpIkTJ+rAgQNKT0/37vu+++6TJDVt2lQtWrRQ\nixYt1LJlSzVu3FhNmjSR0+mUJEVHR2vdunVKTEzUk08+qdjYWA0aNEitW7dW+/btq+0r4DBVDVlq\naPfu3br99tv1008/6dJLL9WYMWP0+9//PlDtAwDUgd/hXlRUpKKiIqWmpurnn39Wenq6Nm7cqJiY\nmEC1EQBQS36XZVq1aqXU1FRJUmxsrJKSkrRu3Tq/GwYAqLuA1tx37NihrVu3Kj09PZC7BQDUUsDC\nvaSkRMOGDdOMGTPUpEmTQO0WAFAHAbmJqbS0VEOGDNGIESM0ePBgn+euvPJK7dy5MxCHAYB6o127\ndn4tEOf3yN0Yo1GjRqlz58566KGHKjy/c+dO7513Nn49/fTTIW8D/TMyBw/K3HefTEGB5/vBg3b1\nz+ZzR/8q/fJ3UOx3uOfm5mrBggX67LPPlJaWprS0NC1btszf3QI1V1wsTZggPf+81Lat5/uECZ7t\nQD3ld1nm17/+tcrKygLRFqBucnM9gf7/N//I6fQ8zs2VzliPBqhPWDjMTy6XK9RNCKqI6F9lAe50\n1ijYI6J/dWRz3yT7++cvv29iOucBzlh4CQBQM/5mJ2vLAICFCHfUP0uWVLzYWlzs2Q5YgnBH/XPN\nNb6zacpn21xzTWjbBQQQNXfUT+WBPn68NH2672wbIAz4m52EO+qvwkIpIUEqKPDMjwfCCBdUgboo\nLvaM2AsKPN+54QmWIdxR/3BHK+oByjKof5Ys8Vw8PbPGXlzMHa0IK9TcAdQf9egfZmruAOoPprHW\nGCN3AJGlnkxjpSwDoP6pB9NYKcsAqF8ibRpriJa7INwBRI5gTmMNVgiH6DoB4Q4gclT3wSz+ClYI\nl7dxwgRPOan8H6cgXyeg5g4A5YJ5sbaW1wmouQNAoDidnmBPSPB8D1Swh+A6AeEOAOWCEcIhWu6C\nsgwASL4h7HRWfFxXdbyrlnnusFM9us0cYSLM/psj3GGnYI2igAgR8guqI0eOVMuWLZWcnOzvrgKP\nz8qMXCGaPgbYwu9wv+uuu7Rs2bJAtCXwWGQosgVr5gJQD/gd7tdee62aN28eiLYEHqO/yBZpt5kD\nYcT+qZCM/iJTJH5aEmVAhJEG5+Mg2dnZ3p9dLpdcLtf5OKzH2aM/Ru6RobrbzMN1tkx5GbCyi8DA\nObjdbrnd7oDtLyCzZQoLCzVw4EBt3ry54gFCOVuGGRc43+rJWuMIvrCYChm24R5m81ZRT9SDtcYR\nfCGfCjl8+HBlZmZq27Ztat26tebNm+fvLgNnwICKoyank2BH8HARGGGCm5iAQKEMiAAKi7JMtQcg\n3FFfUAZEABHuAGChkNfcAQDhh3AHEFrc/BUUhDuA0GINqKCg5g4g9Lj5qwIuqAKwAzd/+eCCKoDI\nx81fAUe4AwitSFwBNAJQlgEQWtz8VSlq7gBgIWruAIAKCHcAsBDhDgAWItwBwEKEOwBYiHAHAAsR\n7pGKlfQiG+cPQUa4RypW0otsnD8EGTcxRTJW0otsnD9UgztU6ztW0otsnD9UgTtU6zNW0otsnD8E\nEeFeF+FwMYyV9CIb5w9B5ne4r1y5UomJiWrfvr1mzZoViDaFv3C4GJab61ujdTo9j3Nzz18bUHec\nPwSZ3zX3tLQ0vfzyy2rTpo369u2rVatWKTY29r8HsLXmzsUwAEHkb3Y28Ofghw4dkiT17NlTkpSV\nlaW8vDwNqA9rMDudnmAvvxhGsAMII36VZb766it17NjR+7hTp05as2aN342KCFwMAxDG/Bq511R2\ndrb3Z5fLJZfLdT4OGzxnXgwrr5We+RgAasntdsvtdgdsf37V3A8dOiSXy6X8/HxJ0gMPPKB+/fr5\nlGWsrLnzsWAAgiyk89wvvvhiSZ4ZM4WFhfrkk0/UvXt3f3YZGQYMqDhCdzoJdoReOEzTRVjweyrk\nzJkz9cc//lE33HCD7rvvPp+ZMgDOs3CYpouwwPIDgG2YpmsF1pYBUFEw1qzhWtN5xdoyAHwFa5pu\nbUo+1P5DjnAHbBLMNWvOnPZbWFj99F9q/yFHWQawyfkondS05EPt3y/U3AGcP7UNbNarrzNq7gDO\nj9qWfFiiI6QYuQOomdqUfM5eouPsxzgnyjIAwg/TJv1GuAOAhai5AwAqINwBwEKEOwBYiHAHAAsR\n7gBgIcIdACxEuAOAhQh3ALAQ4Q4AFiLcAcBChDsAWIhwBwALEe4AYCHCHQAsVOdw//vf/66kpCRd\ncMEFWr9+fSDbhGDg0+iBeqXO4Z6cnKxFixapZ8+egWwPgoVPowfqlTqHe8eOHfWrX/0qkG1BMDmd\n//3My8JCPvIMsBw19/rE6fR8an1Cguc7wR4ZKKmhDhpU92SfPn1UVFRUYfvkyZM1cODAGh8kOzvb\n+7PL5ZLL5arxexFAZ38aPSP3yFBeUqvsw6ZhDbfbLbfbHbgdGj+5XC7z9ddfV/l8AA6BQDh40Jj7\n7vN8r+zx+fC//1vxeAcPerajeuXnq6AgsOeNcxK2/M3OgJRlDB+AHf5yc31H6uU1+Nzc89cGLurW\nXbBKapwTe9X1X4V//OMfJj4+3kRHR5uWLVuafv36Vfo6Pw5R/9SHUVSwRqC2C+bvjXMSlvzNTsf/\n7yRoHA4HI/uaOrOWenZt1abaeGGhZwRaUCC1bRvq1oS/8/HfBeck7PibncyWCSf1Ybri2Rd1z54F\ngoqCXVLjnFiJkXs4snUUVV/+MokknJOwxcjdNjaPosLhoi58cU6sxcg9nITLKGrJEs9siTOPWVzs\n+R9+wIDz1w6gHmPkbpNwGUUxPQ6IeIzcUbnyQB8/nrtZgRDwNzsJd1TN1gu7QASgLIPgsPnCLlAP\nEO6o6MwLuW3b/nfuPQEPRAzKMqiI2TJAyFFzBwALUXMHAFRAuAOAhQh3ALAQ4Q4AFiLcAcBChDsA\nWIhwBwALEe4AYKHQh/uSJRVvay8u9mwHANRJ6MOdtcMBIODCY/kB1g4HAB/2rC3D2uEA4BXStWXG\njx+vxMREXXXVVXrooYd07Nixuu2ItcMBIKD8CvesrCxt3bpV69at05EjR/Tuu+/WfiesHQ4AARew\nssz777+vxYsX6+233/Y9wLn+tGDtcACoIGxq7n379tXdd9+toUOH+h6A9dwBoNb8zc4G53pBnz59\nVFRUVGH75MmTNXDgQEnSM888o5iYmArBXi47O9v7s8vlksvlqltrAcBSbrdbbrc7YPvze+T+5ptv\n6rXXXtM///lPRUdHVzwAI3cAqLWgj9yrs2zZMk2fPl0rV66sNNgBAKHh18i9ffv2OnnypC655BJJ\nUo8ePTR79mzfAzByB4BaC5sLqlUegHAHgFrjA7IBABUQ7gBgIcIdACxEuAOAhQh3ALAQ4Q4AFiLc\nAcBChDsAWIhwBwALEe4AYCHCHQAsRLgDgIUIdwCwEOEOABYi3AHAQoQ7AFiIcAcACxHuAGAhwh0A\nLES4A4CFCHcAsBDhDgAWItwBwEJ1DveJEycqJSVFqampGjFihPbv3x/IdgEA/OAwxpi6vLGkpEQx\nMTGSpGeeeUanTp3SM888U/EADofqeAgAqLf8zc46j9zLg/3UqVM6cuSIoqOj69wIAEBg+VVznzBh\nglq1aqVVq1bpkUceCVSbAAB+qrYs06dPHxUVFVXYPnnyZA0cOFCSdPToUU2YMEGSNGPGjIoHcDj0\n9NNPex+7XC65XC5/2w0AVnG73XK73d7HkyZN8qssU+ea+5k2b96s0aNHa82aNRUPQM0dAGotZDX3\n7du3S/LU3P/2t7/pt7/9bZ0bAQAIrDqH+xNPPKHk5GRlZmbq1KlTGj16dCDbBQDwQ0DKMtUegLIM\nANRayMoyAIDwRbgDgIUIdwCwEOEOABYi3AHAQoQ7AFiIcAcACxHuAGAhwh0ALES4A4CFCHcAsBDh\nDgAWItwBwEKEOwBYiHAHAAsR7gBgIcIdACxEuAOAhQh3ALAQ4Q4AFiLcAcBChDsAWIhwBwAL+R3u\nL730kqKionTgwIFAtAcAEAB+hfvu3bv1ySefqE2bNoFqDwAgAPwK93HjxmnatGmBagsAIEDqHO4f\nfvih4uPj1aVLl0C2B5FmyRKpuNh3W3GxZzuAkGlQ3ZN9+vRRUVFRhe3PP/+8pkyZouXLl3u3GWOq\n3E92drb3Z5fLJZfLVfuWIjxdc400YYL0/POS0+kJ9vLHAGrM7XbL7XYHbH8OU10qV2HLli3q3bu3\nLrroIknSjz/+qMsuu0xr165VixYtfA/gcFQb/LBAeaCPHy9Nn/7foAdQZ/5mZ53C/WwJCQn6+uuv\ndckll1Q8AOFePxQWSgkJUkGB1LZtqFsDRDx/szMg89wdDkcgdoNIVVzsGbEXFHi+n12DB3DeBWTk\nXu0BGLnb7cwa+9k1d0ozQJ2FRVmm2gMQ7nZbssRzUfXMIC8ulnJzpQEDQtcuIMIR7gBgobCouQMA\nwgvhDgAWItwBwEKEOwBYiHAHAAsR7gBgIcIdACxEuAOAhQh3ALAQ4Q4AFiLcAcBChDsAWIhwBwAL\nEe4AYCHCHQAsRLgDgIUIdwCwEOEOABYi3AHAQoQ7AFiozuGenZ2t+Ph4paWlKS0tTcuWLQtkuwAA\nfqhzuDscDo0bN075+fnKz89Xv379AtmuiOF2u0PdhKCif5HL5r5J9vfPX36VZYwxgWpHxLL9PzD6\nF7ls7ptkf//85Ve4z5o1SxkZGZo6dapKSkoC1SYAgJ+qDfc+ffooOTm5wtfixYt17733qqCgQDk5\nOdq5c6deffXV89VmAMC5mADYsGGDyczMrPS5du3aGUl88cUXX3zV4qtdu3Z+5XID1dHevXsVFxen\nU6dO6d1331X//v0rfd2OHTvqeggAQB3Vueb+2GOPqUuXLsrIyFBpaanuvffeQLYLAOAHhzFMeQEA\n2wT1DtWVK1cqMTFR7du316xZs4J5qPOibdu26tKli9LS0pSeni5JKikp0eDBg3X55Zfr5ptv1i+/\n/BLiVtbcyJEj1bJlSyUnJ3u3VdefV155Re3bt1enTp20atWqUDS5Virr39k33y1dutT7XKT1b/fu\n3erVq5eSkpLkcrn07rvvSrLjHFbVN1vO3/Hjx9W9e3elpqYqIyNDM2bMkBTgc+dXxf4cUlNTzYoV\nK0xhYaHp0KGD2bdvXzAPF3Rt27Y1+/fv99k2depUc//995vjx4+bsWPHmunTp4eodbW3cuVKs379\netO5c2fvtqr685///Md06NDB/PDDD8btdpu0tLRQNbvGKutfdna2eemllyq8NhL7t3fvXpOfn2+M\nMWbfvn0mISHBHD582IpzWFXfbDp/R44cMcYYc/z4cZOUlGS2bdsW0HMXtJH7oUOHJEk9e/ZUmzZt\nlJWVpby8vGAd7rwxZ1Wx1q5dq1GjRunCCy/UyJEjI6qP1157rZo3b+6zrar+5OXlqV+/frr88st1\n3XXXyRgT9vc2VNY/qeI5lCKzf61atVJqaqokKTY2VklJSfrqq6+sOIdV9U2y5/xddNFFkqRffvlF\np06d0oUXXhjQcxe0cP/qq6/UsWNH7+NOnTppzZo1wTrceeFwOHT99dfr5ptv1uLFiyX59rNjx45a\nu3ZtKJvot6r6k5eXp8TERO/rOnToELF9rezmu7Vr10Z0/3bs2KGtW7cqPT3dunNY3rfu3btLsuf8\nlZWVKSUlRS1bttT999+vyy+/PKDnjlUhayE3N1cbN27UlClTNG7cOBUVFVm3BENt+uNwOILYkuCo\n6ua7yvodKf0rKSnRsGHDNGPGDDVt2tSqc3hm35o0aWLV+YuKitLGjRu1Y8cOzZ49W/n5+QE9d0EL\n927duum7777zPt66dasyMjKCdbjzIi4uTpKUmJioQYMG6aOPPlK3bt307bffSpK+/fZbdevWLZRN\n9FtV/enevbu++eYb7+u+++67iOxrixYt5HA4dPHFF2vs2LFatGiRpMjtX2lpqYYMGaIRI0Zo8ODB\nkuw5h5X1zbbzJ3kmavTv3195eXkBPXdBC/eLL75YkmfGTGFhoT755BPvn1WR6OjRo94/Afft26ec\nnBz169dP3bt319y5c3Xs2DHNnTs34v8Bq6o/6enpysnJ0a5du+R2uxUVFaWYmJgQt7b29u7dK0kV\nbr6LxP4ZYzRq1Ch17txZDz30kHe7Deewqr7Zcv5+/vlnFRcXS5L279+v5cuXa/DgwYE9dwG9/HsW\nt9ttOnbsaNq1a2defvnlYB4q6L7//nuTkpJiUlJSzPXXX2/eeOMNY4wxhw8fNoMGDTKtW7c2gwcP\nNiUlJSFuac3deuutJi4uzjRq1MjEx8ebuXPnVtufmTNnmnbt2pnExESzcuXKELa8Zsr717BhQxMf\nH2/eeOMNM2LECJOcnGyuvvpq8/DDD/vMfoq0/n3xxRfG4XCYlJQUk5qaalJTU83SpUutOIeV9e3j\njz+25vxt2rTJpKWlmS5dupisrCzz1ltvGWOqz5Pa9o+bmADAQlxQBQALEe4AYCHCHQAsRLgDgIUI\ndwCwEOEOABYi3BEyhw4d0pw5cyR5bk4ZOnRowPa9d+9e9e3bt8avf/XVVzV//vw6H2/y5Ml1fi8Q\nDMxzR8gUFhZq4MCB2rx5c8D3PW/ePBUXF+vhhx8O+L4rExMTE/arEKJ+YeSOkHn88ce1c+dOpaWl\n6ZZbbvF+qMabb76pYcOGKSsrS1dccYXeeustzZkzR126dNHw4cO9Ibpnzx6NHz9ePXr00B133KGC\nggLvvnNycnTjjTfK7Xard+/eGjJkiK688kq98MILWrRokbp27aobb7xRP/74oyTPh0C89NJLkiSX\ny6VJkyapa9euuu6665Sfn+9t1wMPPOA9xk033aQVK1bo8ccf17Fjx5SWlqYRI0ZIkj799FMNHTpU\nPXr08BnVP/7447r66qvVpUsXzZw5M4i/XdR3hDtCZurUqWrXrp3y8/M1ffp0n+dWrlypBQsW6PPP\nP9e9996rAwcOaNOmTWrcuLGWL18uSXrqqad06623avXq1Ro2bJimTZsmSTp9+rT+9a9/eZdOXbVq\nlaZNm6YNGzZo6tSpcrvdWrt2rTIyMrRw4UJJnhX2ylfZczgc+uGHH5SXl6cxY8Z4P0Xs7FX4yt/z\nwgsvqHHjxsrPz9f8+fN19OhRTZ06VfPnz1dubq62bNmivLw8ff/99/ryyy/19ddfa9OmTbrrrruC\n98tFvdcg1A1A/XVmRfDs6uANN9ygFi1aSJKaN2+u4cOHS5J69Oih1atXa/Dgwfr444+1fv36CvvN\ny8vzWaQuPT1d7dq1k+T5XIHBgwcrKipKmZmZ3iVjz3bbbbfpggsuUK9evfTcc89V2saqLF26VN98\n84169OghyfORap9//rnGjRunAwcO6J577tHtt9+uzMzMGu0PqAvCHWHJ6XR6f27UqJH3caNGjXTi\nxAmVlZUpKipKa9as0YUXXujz3qVLl+rGG288574aNmyoEydOVHr88k9watSokY4fPy5Jaty4sc/r\nDxw4UOl7y8rKlJWVpXnz5lV4buPGjfrggw80adIkpaamaurUqVX/EgA/UJZByLRs2VKHDx+u1XvK\nR8+NGjVS//79NWfOHJ0+fVrGGG3atEmS9Nlnn+mGG26o9X7PNTLPyMjQmjVrdPLkSW3ZssXnk3Au\nvfRSHT16VJI0cOBAffHFF951uQ8cOKBdu3Zp//79OnLkiIYMGaLs7OxK/+oAAoWRO0KmcePGGjZs\nmK666ipdeeWVPjXvM+vbZ/9c/njSpEl65ZVX1LVrV508eVLDhw9XXFycoqOj1aRJk0r3daaz6+zV\nvU6S2rRpo4EDByo1NVVJSUlyuVze1zzwwAO69tpr1alTJ82fP1+vvfaaJk6cqG3btqlRo0aaPXu2\noqOjddddd6msrEyxsbF69tln6/ibA86NqZCwyjvvvKM9e/bo0UcfDXVTgJAi3AHAQtTcAcBChDsA\nWIhwBwALEe4AYCHCHQAsRLgDgIUIdwCw0P8B9B9ixDXyHFgAAAAASUVORK5CYII=\n",
       "text": [
        "<matplotlib.figure.Figure at 0x10f91af50>"
       ]
      }
     ],
     "prompt_number": 3
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Our second data set is from ChiP-chip experiments performed on yeast by [Lee et al (2002)](http://www.cs.gsu.edu/~wkim/index_files/ref/TR.pdf). These give us the binding information between transcription factors and genes. In this notebook we are going to try and combine this binding information with the gene expression information to infer transcription factor activities."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "data = GPy.util.datasets.lee_yeast_ChIP()\n",
      "# set S to find relationships where p-value is less than 1e-3\n",
      "S = data['Y'].T<1e-3"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "More details on the data are give in [this notebook](../datasets/leeYeastChip.ipynb)."
     ]
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "Matching the Data"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "The two data sets have slightly different overlapping sets of genes. Fortunately, with `pandas` it is easy to do the matching."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "intersect = list(set(S.columns) & set(Y.columns))\n",
      "# this is the new value for n\n",
      "print 'There are n=', len(intersect), 'overlapping genes.'\n",
      "# Now reset S and Y to have the overlap of genes\n",
      "Y = Y[intersect].T # now an n by T matrix\n",
      "S = S[intersect].T # now an n by q matrix"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "There are n= 6099 overlapping genes.\n"
       ]
      }
     ],
     "prompt_number": 5
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "Model for Transcription Factor Activities"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "We are working with *log* expression levels in a matrix $\\mathbf{Y} \\in \\Re^{n\\times T}$ and we will assume a linear (additive) model giving the relationship between the expression level of the gene and the corresponding transcription factor activity which are unobserved, but we represent by a matrix $\\mathbf{F} \\in \\Re^{q\\times T}$. Our basic assumption is as follows. Transcription factors are in time series, so they are likely to be temporally smooth. Further we assume that the transcription factors are potentially correlated with one another (to account for transcription factors that operate in unison). \n",
      "\n",
      "#### Correlation Between Transcription Factors \n",
      "If there are $q$ transcription factors then correlation between different transcription factors is encoded in a covariance matrix, $\\boldsymbol{\\Sigma}$ which is $q\\times q$ in dimensionality. \n",
      "\n",
      "#### Temporal Smoothness\n",
      "\n",
      "Further we assume that the log of the transcription factors' activities is temporally smooth, and drawn from an underlying Gaussian process with covariance $\\mathbf{K}_t$. \n",
      "\n",
      "#### Intrinsic Coregionalization Model\n",
      "\n",
      "We assume that the joint process across all $q$ transcription factor activities and across all time points is well represented by an *intrinsic model of coregionalization* where the covariance is given by the Kronecker product of these terms.\n",
      "$$\\mathbf{K}_f = \\mathbf{K}_t \\otimes \\boldsymbol{\\Sigma}$$\n",
      "\n",
      "This is known as an intrinsic coregionalization model [Wackernagel, (2003)](http://books.google.co.uk/books/about/Multivariate_Geostatistics.html?id=Rhr7bgLWxx4C). See [Alvarez et al (2012)](http://www.nowpublishers.com/articles/foundations-and-trends-in-machine-learning/MAL-036) for a machine learning orientated review of these methods. The matrix $\\boldsymbol{\\Sigma}$ is known as the coregionalization matrix."
     ]
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "Relation to Gene Expressions"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "We now assume that the $j$th gene's expression is given by the product of the transcription factors that bind to that gene. Because we are working in log space, that implies a log linear relationship. At the $i$th time point, the log of the $j$th gene's expression, $\\mathbf{y}_{i,j}$ is linearly related to the log of the transcription factor activities at the corresponding time point, $\\mathbf{f}_{i, :}$. This relationship is given by the binding information from $\\mathbf{S}$. We then assume that there is some corrupting Gaussian noise to give us the final observation."
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "$$\\mathbf{y}_{i, j} = \\mathbf{S}\\mathbf{f}_{:, i} + \\boldsymbol{\\epsilon}_i$$"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "where the Gaussian noise is sampled from "
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "$$\\epsilon_i \\sim \\mathcal{N}(\\mathbf{0}, \\sigma^2 \\mathbf{I}).$$"
     ]
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "Gaussian Process Model of Gene Expression"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "We consider a vector operator which takes all the separate time series in $\\mathbf{Y}$ and stacks the time series to form a new vector $n\\times T$ length vector $\\mathbf{y}$. A similar operation is applied to form a $q \\times T$ length vector $\\mathbf{f}$. Using Kronecker products we can now represent the relationship between $\\mathbf{y}$ and $\\mathbf{f}$ as follows:  "
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "$$\\mathbf{y} = \\left[\\mathbf{I} \\otimes \\mathbf{S}\\right] \\mathbf{f} + \\boldsymbol{\\epsilon}.$$"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "[Standard properties of multivariate Gaussian distributions](../background/gaussianProperties.ipynb) tell us that "
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "$$\\mathbf{y} \\sim \\mathcal{N}(\\mathbf{0}, \\mathbf{K}),$$"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "where\n",
      "$$\\mathbf{K} = \\mathbf{K}_t \\otimes \\mathbf{S} \\boldsymbol{\\Sigma} \\mathbf{S}^\\top + \\sigma^2 \\mathbf{I}.$$\n",
      "This results in a covariance function that is of size $n$ by $T$ where $n$ is number of genes and $T$ is number of time points. However, we can get a drastic reduction in the size of the covariance function by considering the singular value decomposition of $\\mathbf{S}$. "
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "The matrix $\\mathbf{S}$ is $n$ by $q$ matrix, where $q$ is the number of transcription factors. It contains a 1 if a given transcription factor binds to a given gene, and zero otherwise."
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "$$L = -\\frac{1}{2} \\log |\\mathbf{K}| - \\frac{1}{2} \\mathbf{y}^\\top \\mathbf{K}^{-1} \\mathbf{y}$$"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "In the worst case, because the vector $\\mathbf{y}$ contains $T\\times n$ points ($T$ time points for each of $n$ genes) we are faced with $O(T^3n^3)$ computational complexity. We are going to use a rotation trick to help. "
     ]
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "The Main Computational Trick"
     ]
    },
    {
     "cell_type": "heading",
     "level": 4,
     "metadata": {},
     "source": [
      " Rotating the Basis of a Multivariate Gaussian"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "For any multivariate Gaussian you can rotate the data set and compute a new roated covariance which is valid for the rotated data set. Mathematically this works by first inserting $\\mathbf{R}\\mathbf{R}^\\top$ into the likelihood at three points as follows:"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "$$L = -\\frac{1}{2} \\log |\\mathbf{K}\\mathbf{R}^\\top\\mathbf{R}| - \\frac{1}{2} \\mathbf{y}^\\top\\mathbf{R}^\\top\\mathbf{R} \\mathbf{K}^{-1}\\mathbf{R}^\\top\\mathbf{R} \\mathbf{y} + \\text{const}$$"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "The rules of determinants and a transformation of the data allows us to rewrite the likelihood as"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "$$L = -\\frac{1}{2} \\log |\\mathbf{R}^\\top\\mathbf{K}\\mathbf{R}| - \\frac{1}{2} \\hat{\\mathbf{y}}^\\top \\left[\\mathbf{R}^\\top\\mathbf{K}\\mathbf{R}\\right]^{-1}\\hat{\\mathbf{y}} + \\text{const}$$"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "where we have introduced the rotated data: $\\hat{\\mathbf{y}}=\\mathbf{R} \\mathbf{y}$. Geometrically what this says is that if we want to maintain the same likelihood, then when we rotate our data set by $\\mathbf{R}$ we need to rotate either side of the covariance matrix by $\\mathbf{R}$, which makes perfect sense when we recall the properties of the multivariate Gaussian. "
     ]
    },
    {
     "cell_type": "heading",
     "level": 4,
     "metadata": {},
     "source": [
      "A Kronecker Rotation"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "In this notebook we are using a particular structure of covariance which involves a Kronecker product. The rotation we consider will be a Kronecker rotation (see [Stegle et al, 2011](http://papers.nips.cc/paper/4281-efficient-inference-in-matrix-variate-gaussian-models-with-iid-observation-noise.pdf)). We are going to try and take advantage of the fact that the matrix $\\mathbf{S}$ is square meaning that $\\mathbf{S}\\boldsymbol{\\Sigma}\\mathbf{S}^\\top$ is not full rank (it has rank of most $q$, but is size $n\\times n$, and we expect number of transcription factors $q$ to be less than number of genes $n$). \n",
      "\n",
      "When ranks are involved, it is always a good idea to look at singular value decompositions (SVDs). The SVD of $\\mathbf{S}$ is given by:\n",
      "$$\\mathbf{S} = \\mathbf{Q} \\boldsymbol{\\Lambda} \\mathbf{V}^\\top$$\n",
      "where $\\mathbf{V}^\\top \\mathbf{V} = \\mathbf{I}$, $\\boldsymbol{\\Lambda}$ is a diagonal matrix of positive values, $\\mathbf{Q}$ is a matrix of size $n\\times q$: it matches the dimensionality of $\\mathbf{S}$, but we have $\\mathbf{Q}^\\top \\mathbf{Q} = \\mathbf{I}$. Note that because it is not square, $\\mathbf{Q}$ is not in itself a rotation matrix. However it could be seen as the first $q$ columns of an $n$ dimensional rotation matrix (assuming $n$ is larger than $q$, i.e. there are more genes than transcription factors). \n",
      "\n",
      "If we call the $n-q$ missing columns of this rotation matrix $\\mathbf{U}$ then we have a valid rotation matrix $\\mathbf{R}=\\begin{bmatrix} \\mathbf{Q}& \\mathbf{U}\\end{bmatrix}$. Although this rotation matrix is only rotating across the $n$ dimensions of the genes, not the additional dimensions across time. In other words we are choosing $\\mathbf{K}_t$ to be unrotated. To represent this properly for our covariance we need to set $\\mathbf{R} = \\mathbf{I} \\otimes \\begin{bmatrix} \\mathbf{Q}& \\mathbf{U}\\end{bmatrix}$. This gives us a structure that when applied to a covariance of the form $\\mathbf{K}_t\\otimes \\mathbf{K}_n$ it will rotate $\\mathbf{K}_n$ whilst leaving $\\mathbf{K}_t$ untouched.\n",
      "\n"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "When we apply this rotation matrix to $\\mathbf{K}$ we hace to consider two terms, the rotation of $\\mathbf{K}_t \\otimes \\mathbf{S}\\boldsymbol{\\Sigma}\\mathbf{S}^\\top$, and the rotation of $\\sigma^2 \\mathbf{I}$."
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Rotating the latter is easy, because it is just the identity multiplied by a scalar so it remains unchanged\n",
      "$$\\mathbf{R}^\\top\\mathbf{I}\\sigma^2 \\mathbf{R}= \\mathbf{I}\\sigma^2$$\n",
      "The former is slightly more involved, for that term we have\n",
      "$$\\left[\\mathbf{I}\\otimes \\begin{bmatrix}\\mathbf{Q} & \\mathbf{U}\\end{bmatrix}^\\top \\right]\\mathbf{K}_t \\otimes \\mathbf{S}\\boldsymbol{\\Sigma}\\mathbf{S}^\\top\\left[ \\mathbf{I} \\otimes \\begin{bmatrix}\\mathbf{Q} & \\mathbf{U}\\end{bmatrix}\\right]=\\mathbf{K}_t \\otimes \\begin{bmatrix}\\mathbf{Q} & \\mathbf{U}\\end{bmatrix}^\\top \\mathbf{S} \\boldsymbol{\\Sigma}\\mathbf{S}^\\top \\begin{bmatrix}\\mathbf{Q} & \\mathbf{U}\\end{bmatrix}.$$ \n",
      " "
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Since $\\mathbf{X} = \\mathbf{Q}\\boldsymbol{\\Lambda}\\mathbf{V}^\\top$ then we have\n",
      "$$\\begin{bmatrix}\\mathbf{Q} & \\mathbf{U}\\end{bmatrix}^\\top \\mathbf{X}\\boldsymbol{\\Sigma}\\mathbf{X}^\\top\\begin{bmatrix}\\mathbf{Q} & \\mathbf{U}\\end{bmatrix} = \\begin{bmatrix}\\boldsymbol{\\Lambda} \\mathbf{V}^\\top \\boldsymbol{\\Sigma}\\mathbf{V} \\boldsymbol{\\Lambda} &\\mathbf{0} \\\\ \\mathbf{0} & \\mathbf{0}\\end{bmatrix}.$$\n",
      "\n",
      "This prompts us to split our vector $\\hat{\\mathbf{y}}$ into a $q$ dimensional vector $\\hat{\\mathbf{y}}_u = \\mathbf{U}^\\top \\mathbf{y}$ and an $n-q$ dimensional vector $\\hat{\\mathbf{y}}_q =\\mathbf{Q}^\\top \\mathbf{y}$. The Gaussian likelihood can be written as"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "$$L = L_u + L_q + \\text{const}$$"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "where"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "$$L_q = -\\frac{1}{2} \\log |\\mathbf{K}_t\\otimes\\boldsymbol{\\Lambda}\\mathbf{V}^\\top\\boldsymbol{\\Sigma}\\mathbf{V}\\boldsymbol{\\Lambda}+\\sigma^2\\mathbf{I}| - \\frac{1}{2} \\hat{\\mathbf{y}}_q^\\top \\left[\\mathbf{K}_t\\otimes \\boldsymbol{\\Lambda}\\mathbf{V}^\\top\\boldsymbol{\\Sigma}\\mathbf{V}\\boldsymbol{\\Lambda}+\\sigma^2\\mathbf{I}\\right]^{-1} \\hat{\\mathbf{y}}_q$$"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "and"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "$$L_u = -\\frac{T(n-q)}{2} \\log \\sigma^2  -\\frac{1}{2\\sigma^2} \\hat{\\mathbf{y}}_u^\\top \\hat{\\mathbf{y}}_u$$"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Strictly speaking we should fit these models jointly, but for the purposes of illustration we will firstly use a simple procedure. Firstly, we fit the noise variance $\\sigma^2$ on $\\hat{\\mathbf{y}}_u$ alone using $L_u$. Once this is done, fix the value of $\\sigma^2$ in $L_q$ and optimize with respect to the other parameters."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# step 1, find the SVD of S.\n",
      "n, q = S.shape\n",
      "T = Y.shape[1]\n",
      "R, Lambda, V = np.linalg.svd(S)\n",
      "# Extract first q columns for Q\n",
      "Q = R[:, :q]\n",
      "# remaining columns for U\n",
      "U = R[:, q:]\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 6
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "Q.shape\n",
      "#U.shape\n",
      "#Y.shape"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 7,
       "text": [
        "(6099, 113)"
       ]
      }
     ],
     "prompt_number": 7
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Find sigma2 by looking at variance of y_u\n",
      "Y_u = np.dot(U.T, Y)\n",
      "sigma2 = 1./(T*(n-q))*(Y_u*Y_u).sum()\n",
      "print \"sigma2 found as\", sigma2\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "sigma2 found as 0.178461463478\n"
       ]
      }
     ],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Prepare the data for processing in GPy\n",
      "Y_q = np.dot(Q.T, Y) # project data onto the principal subspace of X \n",
      "\n",
      "# Generate the input associated with each Y, the TF and the time point.\n",
      "x0, x1 = np.asarray(np.meshgrid(t.flatten(),np.arange(q)))\n",
      "X = np.hstack([x0.flatten()[:, None], x1.flatten()[:, None]])\n",
      "y = Y_q.flatten()[:, None]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "\n",
      "kern = GPy.kern.RBF(1, active_dims=[0])*GPy.kern.Coregionalize(1,q,rank=5, active_dims=[1])\n",
      "m = GPy.models.GPRegression(X, y, kern)\n",
      "m.mul.rbf.lengthscale = 50\n",
      "m.Gaussian_noise.variance = sigma2\n",
      "#m.Gaussian_noise.variance.constrain_fixed()\n",
      "#()#kern, t, y_q)\n",
      "#m.sigma2 = constrained_fixed(sigma2)\n",
      "#m.optimize(messages=True)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 50
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print m"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "  GP_regression.           |      Value       |  Constraint  |  Prior  |  Tied to\n",
        "  \u001b[1mmul.rbf.variance       \u001b[0;0m  |             1.0  |     +ve      |         |         \n",
        "  \u001b[1mmul.rbf.lengthscale    \u001b[0;0m  |            50.0  |     +ve      |         |         \n",
        "  \u001b[1mmul.coregion.W         \u001b[0;0m  |        (113, 5)  |              |         |         \n",
        "  \u001b[1mmul.coregion.kappa     \u001b[0;0m  |          (113,)  |     +ve      |         |         \n",
        "  \u001b[1mGaussian_noise.variance\u001b[0;0m  |  0.178461463478  |     +ve      |         |         \n"
       ]
      }
     ],
     "prompt_number": 51
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "m.optimize(messages=True)\n",
      "#m.plot(fixed_inputs=[(1, 1)]) # this would plot ACE2."
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 52
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print m"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "  GP_regression.           |        Value        |  Constraint  |  Prior  |  Tied to\n",
        "  \u001b[1mmul.rbf.variance       \u001b[0;0m  |    0.0114900759537  |     +ve      |         |         \n",
        "  \u001b[1mmul.rbf.lengthscale    \u001b[0;0m  |  5.19983490521e-08  |     +ve      |         |         \n",
        "  \u001b[1mmul.coregion.W         \u001b[0;0m  |           (113, 5)  |              |         |         \n",
        "  \u001b[1mmul.coregion.kappa     \u001b[0;0m  |             (113,)  |     +ve      |         |         \n",
        "  \u001b[1mGaussian_noise.variance\u001b[0;0m  |      0.11622609441  |     +ve      |         |         \n"
       ]
      }
     ],
     "prompt_number": 53
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "With the current design the model is switching off the temporal correlation. The next step in the analysis will be to reimplement the same model as described by [Sanguinetti et al (2006)](http://bioinformatics.oxfordjournals.org/content/22/14/1753.short) and recover their results. That will involve using an Ornstein Uhlbeck covariance and joint maximisation of the likelihoood of $L_u$ and $L_q$."
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "This work is funded by a Commonwealth Scholarship to Aif Rahman and by the EU FP7 RADIANT project."
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": []
    }
   ],
   "metadata": {}
  }
 ]
}